feat(heater-shaker): print subdegree temps

sfoster1 opened the pull request:
We convert and hold all our temperatures in doubles, but weren't
printing them that way, because, well, look at this commit.

We now print them, which requires using snprintf, because nobody's
implemented std::to_chars for float or double.

## Testing
- See if you get nice numbers with decimal points when you send an M105
- No actual conversion code changed anywhere, just the printing code, so that's pretty much it

sfoster1 committed:
feat(heater-shaker): print subdegree temps

We convert and hold all our temperatures in doubles, but weren't
printing them that way, because, well, look at this commit.

We now print them, which requires using snprintf, because nobody's
implemented std::to_chars for float or double.

amitlissack commented on stm32-modules/heater-shaker/include/heater-shaker/gcodes.hpp:
```diff
@@ -119,29 +120,33 @@ struct GetTemperature {
-                            uint32_t setpoint_temperature) -> InputIt {
+                            double current_temperature,
+                            double setpoint_temperature) -> InputIt {
         static constexpr const char* prefix = "M105 C";
```
Pardon the dumb questions....

If you're using snprintf, why not go full on with the format string `"M105 C%02.F T%02.F"`?

What the heck is all this code in the body of a `struct`?

sfoster1 replied:
that's... a good point about the full format string. i probably should do this, yeah.

as to the struct thing - c++ structs are (more or less) classes with default `public` visibility, so these are (static) member functions.

amitlissack replied:
Got it. Thanks

sfoster1 committed:
fixup: just use a big format string

sfoster1 committed:
format

sfoster1 committed:
fixup: be able to print floats

Ah ah ah, you didn't say the magic word! newlib-nano doesn't include the
functionality for printing floats unless you say the magic word!

Laura-Danielle commented on stm32-modules/heater-shaker/include/heater-shaker/gcodes.hpp:
```diff
@@ -69,7 +70,7 @@ struct SetTemperature {
     */
     using ParseResult = std::optional<SetTemperature>;
-    float temperature;
+    double temperature;
```
Does this have to be a double?

Laura-Danielle commented on stm32-modules/heater-shaker/include/heater-shaker/messages.hpp:
```diff
@@ -87,8 +87,8 @@ struct ErrorMessage {
     uint32_t responding_to_id;
-    int16_t current_temperature;
-    uint32_t setpoint_temperature;
+    double current_temperature;
```
same here on the double, is it really necessary? I know this type is generally more expensive than others.

sfoster1 replied:
Doubles aren't really truly necessary, but it's nice to just not have to worry about it. And the timescales for the temperature stuff is so much slower than anything else we care about that it'll be fine.

Laura-Danielle reviewed (approved):
Just a comment about doubles, otherwise looks good.
